File: apps/android/core/src/main/kotlin/ai/deepseek/dsh/link/LinkWire.kt
package ai.deepseek.dsh.link

import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * A JSON value carried straight through either envelope direction — the
 * Kotlin mirror of the Swift `LinkWire.WireValue` and the TypeScript
 * reference client's pass-through JSON: encoded into request payloads,
 * decoded out of response results and stream frames.
 */
sealed class WireValue {
    data class StringValue(val value: String) : WireValue()

    data class NumberValue(val value: Double) : WireValue()

    data class BoolValue(val value: Boolean) : WireValue()

    data object NullValue : WireValue()

    data class ArrayValue(val items: List<WireValue>) : WireValue()

    data class ObjectValue(val entries: Map<String, WireValue>) : WireValue()

    /** Canonical JSON tree form for encoding. */
    fun toJsonElement(): JsonElement = when (this) {
        is StringValue -> JsonPrimitive(value)
        is NumberValue -> if (value == kotlin.math.floor(value) && !value.isInfinite()) JsonPrimitive(value.toLong()) else JsonPrimitive(value)
        is BoolValue -> JsonPrimitive(value)
        is NullValue -> JsonNull
        is ArrayValue -> JsonArray(items.map { it.toJsonElement() })
        is ObjectValue -> buildJsonObject {
            for ((key, inner) in entries) put(key, inner.toJsonElement())
        }
    }

    companion object {
        /** Parse one JSON tree into the pass-through value. */
        fun fromJsonElement(element: JsonElement): WireValue = when (element) {
            is JsonPrimitive -> when {
                element.isString -> StringValue(element.content)
                element.booleanOrNull != null -> BoolValue(element.booleanOrNull!!)
                element.contentOrNullSafe == "null" -> NullValue
                else -> {
                    val number = element.content.toDoubleOrNull() ?: invalidResponse()
                    if (!number.isFinite() || number == 0.0 && number.toRawBits() < 0) invalidResponse()
                    NumberValue(number)
                }
            }
            is JsonArray -> ArrayValue(element.map { fromJsonElement(it) })
            is JsonObject -> ObjectValue(element.entries.associate { (key, inner) -> key to fromJsonElement(inner) })
            JsonNull -> NullValue
        }
    }
}

private val JsonPrimitive.contentOrNullSafe: String?
    get() = content

/** The unary request envelope: `{ type, rpcId, method, payload: { args, device? } }`.
 * The optional `device` admission rides beside `args`, mirroring the gateway's
 * versioned request envelope; a paired client sends it on every business call. */
data class LinkRequestEnvelope(
    val rpcId: String,
    val method: String,
    val args: Map<String, WireValue>,
    val device: Map<String, WireValue>? = null,
) {
    fun toJsonElement(): JsonElement = buildJsonObject {
        put("type", "client-request")
        put("rpcId", rpcId)
        put("method", method)
        put("payload", buildJsonObject {
            put("args", buildJsonObject { args.forEach { (key, value) -> put(key, value.toJsonElement()) } })
            device?.let { admission ->
                put("device", buildJsonObject { admission.forEach { (key, value) -> put(key, value.toJsonElement()) } })
            }
        })
    }
}

/** `{ ok: true, value }` or `{ ok: false, error }` — one unary result. */
data class LinkResult(
    val ok: Boolean,
    val value: WireValue?,
    val errorCode: String?,
    val errorMessage: String?,
    val errorDetails: WireValue?,
) {
    companion object {
        fun fromJsonElement(element: JsonElement): LinkResult {
            val obj = element.responseObject()
            val ok = (obj["ok"] as? JsonPrimitive)?.takeUnless { it.isString }?.booleanOrNull ?: invalidResponse()
            if (ok) {
                obj.requireAbsent("error")
                return LinkResult(true, obj["value"]?.let { WireValue.fromJsonElement(it) }, null, null, null)
            }
            obj.requireAbsent("value")
            val error = obj["error"].responseObject()
            return LinkResult(
                false,
                null,
                error.responseString("code"),
                error.responseString("message"),
                WireValue.fromJsonElement(error["details"].responseObject()),
            )
        }
    }
}

/** One whole server response: `{ type, rpcId, result }`. */
data class LinkResponseEnvelope(val type: String, val rpcId: String, val result: LinkResult) {
    companion object {
        fun fromJsonElement(element: JsonElement): LinkResponseEnvelope {
            val obj = element.responseObject()
            val type = obj.responseString("type")
            val rpcId = obj.responseString("rpcId")
            if (type != "server-response" || rpcId.isEmpty()) invalidResponse()
            return LinkResponseEnvelope(
                type = type,
                rpcId = rpcId,
                result = LinkResult.fromJsonElement(obj["result"].responseObject()),
            )
        }
    }
}

/** One NDJSON Remote-stream frame: `{"k":"v","v":…}` or `{"k":"e",…}`. */
data class DecodedLinkStreamFrame(
    val kind: String,
    val value: WireValue?,
    val code: String?,
    val message: String?,
    val details: WireValue?,
) {
    val isFailure: Boolean get() = kind == "e"

    companion object {
        fun fromJsonElement(element: JsonElement): DecodedLinkStreamFrame {
            val obj = element.responseObject()
            return when (obj.responseString("k")) {
                "v" -> {
                    obj.requireAbsent("c", "m", "d")
                    DecodedLinkStreamFrame("v", obj["v"]?.let { WireValue.fromJsonElement(it) }, null, null, null)
                }
                "e" -> {
                    obj.requireAbsent("v")
                    DecodedLinkStreamFrame(
                        "e",
                        null,
                        obj.responseString("c"),
                        obj.responseString("m"),
                        WireValue.fromJsonElement(obj["d"].responseObject()),
                    )
                }
                else -> invalidResponse()
            }
        }
    }
}

/** Decode response bytes without exposing a JSON parser exception as a transport failure. */
internal fun responseJson(text: String): JsonElement = try {
    kotlinx.serialization.json.Json.parseToJsonElement(text)
} catch (_: IllegalArgumentException) {
    invalidResponse()
}

private fun JsonElement?.responseObject(): JsonObject = this as? JsonObject ?: invalidResponse()

private fun JsonObject.responseString(field: String): String =
    (this[field] as? JsonPrimitive)?.takeIf { it.isString }?.content ?: invalidResponse()

private fun JsonObject.requireAbsent(vararg fields: String) {
    if (fields.any { containsKey(it) }) invalidResponse()
}

private fun invalidResponse(): Nothing = throw LinkClientException.BadWire("invalid response fields")

File: apps/apple/Sources/SharedAppleRemoteCore/LinkWire.swift
import Foundation

/// The Remote wire envelope: exactly what the gateway's shared `/api` chain
/// expects, decoded from the TypeScript reference client byte-for-byte.
public enum LinkWire {

    /// One unary request body: `{ type, rpcId, method, payload: { args } }`.
    public struct RequestEnvelope: Encodable {
        public let type = "client-request"
        public let rpcId: String
        public let method: String
        public let payload: Payload

        public struct Payload: Encodable {
            public let args: [String: Value]

            /// A JSON value encoded straight through.
            public enum Value: Encodable {
                case string(String)
                case number(Double)
                case bool(Bool)
                case null
                case array([Value])
                case object([String: Value])

                public func encode(to encoder: Encoder) throws {
                    var container = encoder.singleValueContainer()
                    switch self {
                    case .string(let value): try container.encode(value)
                    case .number(let value): try container.encode(value)
                    case .bool(let value): try container.encode(value)
                    case .null: try container.encodeNil()
                    case .array(let values): try container.encode(values)
                    case .object(let entries): try container.encode(entries)
                    }
                }
            }

            public init(args: [String: Value]) {
                self.args = args
            }
        }

        public init(rpcId: String, method: String, args: [String: Payload.Value]) {
            self.rpcId = rpcId
            self.method = method
            self.payload = Payload(args: args)
        }
    }

    /// One unary response body: `{ type: "server-response", rpcId, result }`.
    public struct ResponseEnvelope: Decodable {
        public let type: String
        public let rpcId: String
        public let result: Result

        /// `{ ok: true, value }` or `{ ok: false, error }`.
        public struct Result: Decodable {
            public let ok: Bool
            public let value: Value?
            public let error: Failure?

            /// A JSON value decoded straight through.
            public enum Value: Decodable, Equatable {
                case string(String)
                case number(Double)
                case bool(Bool)
                case null
                case array([Value])
                case object([String: Value])

                public init(from decoder: Decoder) throws {
                    let container = try decoder.singleValueContainer()
                    if container.decodeNil() { self = .null; return }
                    if let value = try? container.decode(Bool.self) { self = .bool(value); return }
                    if let value = try? container.decode(Double.self) { self = .number(value); return }
                    if let value = try? container.decode(String.self) { self = .string(value); return }
                    if let value = try? container.decode([Value].self) { self = .array(value); return }
                    if let value = try? container.decode([String: Value].self) { self = .object(value); return }
                    throw DecodingError.dataCorruptedError(in: container, debugDescription: "unsupported JSON value")
                }
            }

            /// `{ code, message, details }` on a refused call.
            public struct Failure: Decodable, Equatable {
                public let code: String
                public let message: String
            }
        }
    }

    /// One NDJSON Remote-stream frame: `{"k":"v","v":…}` or `{"k":"e",…}`.
    public struct StreamFrame: Decodable {
        public let k: String
        public let v: ResponseEnvelope.Result.Value?
        public let c: String?
        public let m: String?

        /// `true` when the frame carries a stream value.
        public var isValue: Bool { k == "v" }
        /// `true` when the frame terminates the stream with a typed failure.
        public var isFailure: Bool { k == "e" }
    }
}
